base/range.jl

```diff
@@ -275,6 +275,10 @@ RangeStepStyle(::Type{<:AbstractRange{<:Integer}}) = RangeStepRegular()
 
 convert(::Type{T}, r::AbstractRange) where {T<:AbstractRange} = r isa T ? r : T(r)::T
 
+AbstractRange{T}(r::AbstractRange) where T = T(first(r)):T(step(r)):T(last(r))
+AbstractArray{T,1}(r::AbstractRange) where T = AbstractRange{T}(r)
+AbstractArray{T}(r::AbstractRange) where T = AbstractRange{T}(r)
```

Comment:
Here are some failures from this `T(first(r)):T(step(r)):T(last(r))` construction:
```julia
julia> convert(AbstractArray{Float32}, 0 * range(1, 2, 10)) |> length
ERROR: ArgumentError: range step cannot be zero

julia> convert(AbstractArray{Float16}, range(1/43^2, 1, 43)) |> length
42
```

Reply:
Thanks for pointing that out!

`range` would fix `StepRangeLen`, but changes `OrdinalRange` to `StepRangeLen` as well. Anyway, either `OrdinalRange` or `StepRangeLen` needs a specialized method.

As for the general fallback, I would keep the colon syntax until there are other issues.

Reply:
Wait the priority surely is to get the right answer?

If necessary, there can be specialised paths which try also to preserve certain types, like UnitRange and StepRange and LinRange, when possible. But the default path should not rely on `a:b:c` colon magic to keep away from factually wrong answers.

Reply:
It is not a matter of correctness, since specializing `StepRangeLen` solves the problem. Colon or `range(start, step, stop)` suits `OrdinalRange` better, and the implementations within the subtype tree of `OrdinalRange` is significantly more sophiscated.

Reply:
My point is that the abstract fallback should try very hard to be correct. The goal isn't just to cover the types in Base with the smallest number of methods. 

Someone, somewhere, is going to have their own `MyRange <: AbstractRange` whose `step` is a floating point number. That type will run into the problem I pointed out. It's generic to floating point, not specific to StepRangeLen.

Reply:
The same goes for ranges that use `step` over `length`, for example, `InfiniteArrays.InfStepRange` (offtopic note: it's also not integer-exclusive). It ultimately depends on which choice is (or is predicted to be) more popular on custom packages, and I'm not yet convinced that more packages have custom floating point ranges than integer ones.

Reply:
I'm not sure I follow. Preserving integer steps exactly is a nice-to-have, which you can get by specialising on OrdinalRange. But preserving length of a vector, always, is not optional.

I know nothing about infinite arrays, and am not sure they are fully legal AbstractArrays. But the type you mention seems to be `InfiniteArrays.InfStepRange{T, S} <: OrdinalRange{T, S} <: AbstractRange{T} <: AbstractArray{T, 1} <: Any`, so it would hit a specialisation on OrdinalRange.